File: go/055.go
func canJump(nums []int) bool {
    start, end, size := 0, 0, len(nums)
    for start < size {
        next := 0
        for i := start; i <= end; i++ {
            next = max(next, i+nums[i])
            if next >= size-1 {
                return true
            }
        }
        if next <= end {
            return false
        }
        start, end = end+1, next
    }
    return true
}

func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

File: cpp/055.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int start = 0, end = 0, size= nums.size();
        for (; start < size; ) {
            int next = 0;
            for (int i = start; i <= end; ++i) {
                next = max(next, i + nums[i]);
                if (next >= size - 1)
                    return true;
            }
            if (next <= end)
                return false;
            start = end + 1;
            end = next;
        }
        return true;
    }
};
